### PR TITLE
LSP: Surround With refactorings implemented.

### PR DESCRIPTION
**Pull request by dbalek**
Surround With refactorings available via LSP.

### CODE REVIEW

File: java/java.lsp.server/src/org/netbeans/modules/java/lsp/server/protocol/SurroundWithHint.java [hunk 1]
```diff
@@ -0,0 +1,298 @@
...
+    public CompletableFuture<Object> processCommand(NbCodeLanguageClient client, String command, List<Object> arguments) {
+        return CompletableFuture.completedFuture(false);
+    }
+
+    private static boolean isRightParenthesisOfLoopPresent(CompilationInfo info, int abbrevStartOffset) {
+        TokenHierarchy<?> tokenHierarchy = info.getTokenHierarchy();
+        TokenSequence<?> tokenSequence = tokenHierarchy.tokenSequence();
+        tokenSequence.move(abbrevStartOffset);
+        if (tokenSequence.moveNext()) {
+            TokenId tokenId = skipNextWhitespaces(tokenSequence);
```
Comment (sdedic) [inline java/java.lsp.server/src/org/netbeans/modules/java/lsp/server/protocol/SurroundWithHint.java:191]: There are two `moveNext()` from the `abbrevStartOffset`, so the token at offset itself is skipped. OK ?

File: java/java.lsp.server/src/org/netbeans/modules/java/lsp/server/protocol/SurroundWithHint.java [hunk 2]
```diff
@@ -0,0 +1,298 @@
...
+                    String headerText = info.getText().substring(startPos, startOffset);
+                    int idx = headerText.indexOf('{'); //NOI18N
+                    if (idx < 0) {
+                        treeKindCtx = null;
+                        stringCtx = CLASS_HEADER;
+                    }
+                    break;
+                case FOR_LOOP:
+                case ENHANCED_FOR_LOOP:
+                    if (!isRightParenthesisOfLoopPresent(info, startOffset)) {
```
Comment (sdedic) [inline java/java.lsp.server/src/org/netbeans/modules/java/lsp/server/protocol/SurroundWithHint.java:144]: Maybe missing a testcase for this TreeKind...

File: java/java.lsp.server/src/org/netbeans/modules/java/lsp/server/protocol/SurroundWithHint.java [hunk 3]
```diff
@@ -0,0 +1,298 @@
...
+            if (parametrizedText.toLowerCase().contains(SELECTION_VAR) && !TO_FILTER.contains(codeTemplate.getAbbreviation())) {
+                if (codeTemplate.getContexts() == null || codeTemplate.getContexts().isEmpty() || accept(codeTemplate, treeKindCtx, stringCtx)) {
+                    String label = html2text(codeTemplate.getDescription());
+                    if (label == null) {
+                        String text = codeTemplate.getParametrizedText();
+                        int idx = text.indexOf('\n');
+                        label = convert(idx < 0 ? text : text.substring(0, idx) + DOTS, true);
+                    }
+                    String snippet = convert(codeTemplate.getParametrizedText(), false);
+                    codeActions.add(createCodeAction(Bundle.DN_SurroundWith() + label, CodeActionKind.RefactorRewrite, COMMAND_INSERT_SNIPPET, Collections.singletonMap(SNIPPET, snippet)));
```
[the remark this answers is not in the document]
Comment (dbalek) [reply java/java.lsp.server/src/org/netbeans/modules/java/lsp/server/protocol/SurroundWithHint.java:169]: Fixed.

File: java/java.lsp.server/src/org/netbeans/modules/java/lsp/server/protocol/SurroundWithHint.java [hunk 4]
```diff
@@ -0,0 +1,298 @@
...
+                    }
+                    break;
+                case CLASS:
+                    SourcePositions sp = info.getTrees().getSourcePositions();
+                    int startPos = (int)sp.getEndPosition(info.getCompilationUnit(), ((ClassTree)tree).getModifiers());
+                    if (startPos <= 0) {
+                        startPos = (int)sp.getStartPosition(info.getCompilationUnit(), tree);
+                    }
+                    String headerText = info.getText().substring(startPos, startOffset);
+                    int idx = headerText.indexOf('{'); //NOI18N
```
[the remark this answers is not in the document]
Comment (dbalek) [reply java/java.lsp.server/src/org/netbeans/modules/java/lsp/server/protocol/SurroundWithHint.java:136]: In this case, `startPos` is placed at the end of class modifiers that include also class annotations (see `ModifiersTree.getAnnotations()`). So the class' starting curly brace identification should be OK.

File: java/java.lsp.server/src/org/netbeans/modules/java/lsp/server/protocol/SurroundWithHint.java (hunk 1 above)
Comment (dbalek) [reply java/java.lsp.server/src/org/netbeans/modules/java/lsp/server/protocol/SurroundWithHint.java:191]: From the `TokenSequence.move(int offset)` Javadoc: "Move token sequence to be positioned between `index-1`
and `index` tokens. .... Subsequent `moveNext()` or `movePrevious()` is needed to fetch a concrete token..."

File: ide/editor.codetemplates/apichanges.xml [hunk 5]
```diff
@@ -84,6 +84,18 @@ is the proper place.
...
     <changes>
+        <change id="CodeTemplateFilter.Factory.createForDoc">
+             <api name="codetemplates"/>
+             <summary>Default method <code>createFilter(Document, int, int)</code> added to <code>CodeTemplateFilter.Factory</code>.</summary>
+             <version major="1" minor="57"/>
+             <date day="9" month="9" year="2021"/>
+             <author login="dbalek"/>
+             <compatibility addition="yes" binary="compatible" deletion="no" deprecation="no" modification="no" semantic="compatible" source="compatible"/>
+             <description>
+                 Added method allows for creating code template filters for the given document and range.
```
Comment (JaroslavTulach) [inline ide/editor.codetemplates/apichanges.xml:95]: "allows for"? Isn't just "allows creating" better?

**Review by JaroslavTulach — APPROVED** [review]

File: ide/editor.codetemplates/apichanges.xml (hunk 5 above)
Comment (dbalek) [reply ide/editor.codetemplates/apichanges.xml]: Fixed.

File: java/java.lsp.server/src/org/netbeans/modules/java/lsp/server/protocol/SurroundWithHint.java (hunk 2 above)
Comment (dbalek) [reply java/java.lsp.server/src/org/netbeans/modules/java/lsp/server/protocol/SurroundWithHint.java]: Replaced with a call to the existing `CodeTemplateFilter`s.

**Pull request merged**